### PR TITLE
Change the type of parser "extract" function to "Step"

### LINKED ISSUE #604: Bug in alternative instance of ParserD
ParserD's `<|>` is not trying the other parser in case the first parser fails in the `extract` function, this behaviour is different from ParserK's `<|>` implementation.

### PR DESCRIPTION
**Pull request by harendra-kumar**
fixes #604

### CODE REVIEW

**Comment by harendra-kumar** [comment]
Supersedes #615 .

**Comment by harendra-kumar** [comment]
Benchmarks:

```
Data.Parser.ParserD(cpuTime)
Benchmark                                                                      default(0)(μs) default(1) - default(0)(%)
------------------------------------------------------------------------------ -------------- --------------------------
All.Data.Parser.ParserD/o-n-heap.lookAhead                                              37.37                   +3521.15
All.Data.Parser.ParserD/o-1-space.parseMany groupRollingByEither (Left)                 84.06                    +199.04
All.Data.Parser.ParserD/o-1-space.parseMany groupRollingByEither (Right)                87.30                    +188.27
All.Data.Parser.ParserD/o-1-space.some                                                  37.68                    +163.79
All.Data.Parser.ParserD/o-1-space.sequenceParser                                      4311.21                     +10.27
All.Data.Parser.ParserD/o-1-space.parseMany/Unfold/1000 arrays/100000 parses           330.30                      +3.18

Data.Parser(cpuTime)
Benchmark                                                                   default(0)(μs) default(1) - default(0)(%)
--------------------------------------------------------------------------- -------------- --------------------------
All.Data.Parser/o-1-space.takeEQ                                                     37.32                   +3520.94
All.Data.Parser/o-1-space.some                                                       37.68                    +164.14
All.Data.Parser/o-n-heap.someAlt                                                  25817.40                     +25.80
All.Data.Parser/o-1-space.parseBreak (recursive)                                   9485.14                      +7.61
All.Data.Parser/o-1-space.splitAp                                                  1313.90                      +6.90
All.Data.Parser/o-1-space.splitApAfter                                             1331.14                      +3.23
```

File: core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs [hunk 1]
```diff
@@ -378,35 +374,26 @@ instance Functor (Step s) where
...
 --
 {-# INLINE extractStep #-}
 extractStep :: Monad m => (s -> m (Step s1 b)) -> Step s b -> m (Step s1 b)
 extractStep f res =
     case res of
-        -- XXX When step says (Partial n) extract must say (Done n)
-        -- We can assert that.
         Partial n s1 -> assertStepCount n <$> f s1
         Done n b -> return $ assertStepCount n $ Done n b
-        -- XXX What to do with n?
```
Comment (adithyaov) [inline core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs]: You can remove `assertStepCount` in the above line.

Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Parser/ParserD/Type.hs]: Ah, yes. That's redundant.

**Pull request merged**